File: backend/app/api/v1/endpoints/webhooks.py
"""Webhook / Event Engine endpoints."""
from __future__ import annotations
from typing import Any, Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Body, Depends, Header, HTTPException, Query, Request, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.core.deps import get_current_user, get_current_active_superuser
from app.models.webhook import DeliveryStatus
from app.schemas.webhook import (
    DeliveryAttemptOut, EventDefinitionCreate, EventDefinitionOut,
    EventLogOut, InboundEndpointCreate, InboundEndpointOut,
    PublishEventRequest, ReplayRequest, RetryRequest,
    SubscriptionCreate, SubscriptionOut, SubscriptionUpdate,
    WebhookDashboard, WebhookReport,
)
import app.services.webhook_service as svc

router = APIRouter()


# ── Dashboard ─────────────────────────────────────────────────────────────────

@router.get("/dashboard", response_model=WebhookDashboard)
async def dashboard(db: AsyncSession = Depends(get_db)):
    return await svc.get_dashboard(db)


# ── Event Definitions ─────────────────────────────────────────────────────────

@router.get("/event-definitions", response_model=List[EventDefinitionOut])
async def list_event_definitions(
    module: Optional[str] = None,
    active_only: bool = True,
    skip: int = 0,
    limit: int = Query(default=100, le=500),
    db: AsyncSession = Depends(get_db),
):
    return await svc.list_event_definitions(db, module=module, active_only=active_only, skip=skip, limit=limit)


@router.post("/event-definitions", response_model=EventDefinitionOut, status_code=201)
async def create_event_definition(
    data: EventDefinitionCreate,
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_active_superuser),
):
    return await svc.create_event_definition(db, data)


@router.post("/event-definitions/seed", status_code=200)
async def seed_event_definitions(
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_active_superuser),
):
    await svc.seed_event_definitions(db)
    return {"detail": "Standard event definitions seeded"}


# ── Event Logs ────────────────────────────────────────────────────────────────

@router.get("/events", response_model=List[EventLogOut])
async def list_events(
    event_code: Optional[str] = None,
    module: Optional[str] = None,
    reference_type: Optional[str] = None,
    reference_id: Optional[str] = None,
    skip: int = 0,
    limit: int = Query(default=100, le=500),
    db: AsyncSession = Depends(get_db),
):
    from sqlalchemy import select, desc, and_
    from app.models.webhook import EventLog
    q = select(EventLog).order_by(desc(EventLog.created_at))
    if event_code:
        q = q.where(EventLog.event_code == event_code)
    if module:
        q = q.where(EventLog.module == module)
    if reference_type:
        q = q.where(EventLog.reference_type == reference_type)
    if reference_id:
        q = q.where(EventLog.reference_id == reference_id)
    result = await db.execute(q.offset(skip).limit(limit))
    return list(result.scalars().all())


@router.get("/events/{event_id}", response_model=EventLogOut)
async def get_event(event_id: UUID, db: AsyncSession = Depends(get_db)):
    from sqlalchemy import select
    from app.models.webhook import EventLog
    result = await db.execute(select(EventLog).where(EventLog.id == event_id))
    obj = result.scalar_one_or_none()
    if not obj:
        raise HTTPException(404, "Event not found")
    return obj


@router.post("/events/publish", response_model=EventLogOut, status_code=201)
async def publish_event(data: PublishEventRequest, db: AsyncSession = Depends(get_db)):
    """Publish an event manually (or from internal module calls)."""
    return await svc.publish_event(
        db,
        data.event_code,
        data.payload,
        reference_type=data.reference_type,
        reference_id=data.reference_id,
        actor_id=data.actor_id,
        partition_key=data.partition_key,
    )


@router.post("/events/replay")
async def replay_events(
    data: ReplayRequest,
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_active_superuser),
):
    count = await svc.replay_events(db, data)
    return {"detail": f"Queued {count} delivery attempt(s)"}


# ── Subscriptions ─────────────────────────────────────────────────────────────

@router.get("/subscriptions", response_model=List[SubscriptionOut])
async def list_subscriptions(
    active_only: bool = False,
    skip: int = 0,
    limit: int = Query(default=100, le=500),
    db: AsyncSession = Depends(get_db),
):
    return await svc.list_subscriptions(db, active_only=active_only, skip=skip, limit=limit)


@router.post("/subscriptions", response_model=SubscriptionOut, status_code=201)
async def create_subscription(
    data: SubscriptionCreate,
    db: AsyncSession = Depends(get_db),
):
    return await svc.create_subscription(db, data)


@router.get("/subscriptions/{sub_id}", response_model=SubscriptionOut)
async def get_subscription(sub_id: UUID, db: AsyncSession = Depends(get_db)):
    obj = await svc.get_subscription(db, sub_id)
    if not obj:
        raise HTTPException(404, "Subscription not found")
    return obj


@router.patch("/subscriptions/{sub_id}", response_model=SubscriptionOut)
async def update_subscription(sub_id: UUID, data: SubscriptionUpdate, db: AsyncSession = Depends(get_db)):
    obj = await svc.update_subscription(db, sub_id, data)
    if not obj:
        raise HTTPException(404, "Subscription not found")
    return obj


@router.delete("/subscriptions/{sub_id}", status_code=204)
async def delete_subscription(sub_id: UUID, db: AsyncSession = Depends(get_db)):
    ok = await svc.delete_subscription(db, sub_id)
    if not ok:
        raise HTTPException(404, "Subscription not found")


@router.post("/subscriptions/{sub_id}/test")
async def test_subscription(sub_id: UUID, db: AsyncSession = Depends(get_db)):
    return await svc.test_subscription(db, sub_id)


# ── Delivery Attempts ─────────────────────────────────────────────────────────

@router.get("/deliveries", response_model=List[DeliveryAttemptOut])
async def list_deliveries(
    subscription_id: Optional[UUID] = None,
    event_log_id: Optional[UUID] = None,
    delivery_status: Optional[DeliveryStatus] = Query(default=None),
    skip: int = 0,
    limit: int = Query(default=100, le=500),
    db: AsyncSession = Depends(get_db),
):
    return await svc.list_deliveries(
        db,
        subscription_id=subscription_id,
        event_log_id=event_log_id,
        status=delivery_status,
        skip=skip,
        limit=limit,
    )


@router.get("/deliveries/dead-letter", response_model=List[DeliveryAttemptOut])
async def list_dead_letter(
    skip: int = 0,
    limit: int = Query(default=100, le=500),
    db: AsyncSession = Depends(get_db),
):
    return await svc.list_deliveries(db, status=DeliveryStatus.DEAD_LETTER, skip=skip, limit=limit)


@router.post("/deliveries/{delivery_id}/retry", response_model=DeliveryAttemptOut)
async def retry_delivery(delivery_id: UUID, db: AsyncSession = Depends(get_db)):
    obj = await svc.manual_retry_delivery(db, delivery_id)
    if not obj:
        raise HTTPException(404, "Delivery not found")
    return obj


@router.post("/deliveries/process-pending")
async def process_pending(
    limit: int = 50,
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_active_superuser),
):
    """Trigger processing of pending delivery queue (background sweep)."""
    processed = await svc.process_pending_deliveries(db, limit=limit)
    retried = await svc.process_retry_deliveries(db, limit=limit)
    return {"processed": processed, "retried": retried}


# ── Inbound Endpoints ─────────────────────────────────────────────────────────

@router.get("/inbound-endpoints", response_model=List[InboundEndpointOut])
async def list_inbound_endpoints(db: AsyncSession = Depends(get_db)):
    return await svc.list_inbound_endpoints(db)


@router.post("/inbound-endpoints", response_model=InboundEndpointOut, status_code=201)
async def create_inbound_endpoint(
    data: InboundEndpointCreate,
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_active_superuser),
):
    return await svc.create_inbound_endpoint(db, data)


@router.api_route("/inbound/{endpoint_path:path}", methods=["POST", "PUT", "GET"])
async def receive_inbound_webhook(
    endpoint_path: str,
    request: Request,
    db: AsyncSession = Depends(get_db),
):
    """Dynamic inbound webhook receiver. Validates auth and logs event."""
    endpoint = await svc.get_inbound_endpoint_by_path(db, endpoint_path)
    if not endpoint:
        raise HTTPException(404, f"Inbound endpoint '{endpoint_path}' not found")

    body = await request.body()
    sig_header = request.headers.get("X-Webhook-Signature") or request.headers.get("Authorization")

    if not svc.verify_inbound_signature(endpoint, body, sig_header):
        raise HTTPException(status_code=401, detail="Invalid webhook signature")

    try:
        import json
        payload = json.loads(body) if body else {}
    except Exception:
        payload = {"raw_body": body.decode(errors="replace")[:5000]}

    # Log as event
    await svc.publish_event(
        db,
        event_code=f"inbound.{endpoint.handler_module or endpoint_path}",
        payload=payload,
        reference_type="inbound_endpoint",
        reference_id=endpoint_path,
        module="inbound",
    )

    # Update receive counter
    endpoint.receive_count = (endpoint.receive_count or 0) + 1
    from datetime import datetime, timezone
    endpoint.last_received_at = datetime.now(timezone.utc)
    await db.commit()

    return {"received": True, "endpoint": endpoint_path}


# ── Reports ───────────────────────────────────────────────────────────────────

@router.get("/reports", response_model=WebhookReport)
async def get_report(
    days: int = Query(default=7, ge=1, le=90),
    db: AsyncSession = Depends(get_db),
):
    return await svc.get_report(db, days=days)


# ── AI Agents ─────────────────────────────────────────────────────────────────

@router.get("/ai/health-monitor")
async def ai_health_monitor(
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    return await svc.ai_health_monitor(db)


@router.get("/ai/payload-advisor")
async def ai_payload_advisor(
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    return await svc.ai_payload_advisor(db)


@router.get("/ai/anomaly-detector")
async def ai_anomaly_detector(
    db: AsyncSession = Depends(get_db),
    _=Depends(get_current_user),
):
    return await svc.ai_anomaly_detector(db)

File: frontend/src/lib/webhooks.ts
import { apiClient } from "./api";

export type HttpMethod = "POST" | "PUT";
export type AuthType = "none" | "basic" | "bearer" | "hmac";
export type DeliveryStatus = "pending" | "sent" | "failed" | "retrying" | "dead_letter";

export interface EventDefinition {
  id: string;
  event_code: string;
  event_name: string;
  module: string;
  description?: string;
  active_flag: boolean;
  created_at: string;
}

export interface EventLog {
  id: string;
  event_code: string;
  module: string;
  reference_type?: string;
  reference_id?: string;
  payload: Record<string, unknown>;
  actor_id?: string;
  partition_key?: string;
  created_at: string;
}

export interface Subscription {
  id: string;
  name: string;
  endpoint_url: string;
  http_method: HttpMethod;
  auth_type: AuthType;
  event_codes?: string[];
  filter_expression?: Array<{ field: string; op: string; value: unknown }>;
  max_retries: number;
  timeout_seconds: number;
  active_flag: boolean;
  created_at: string;
}

export interface SubscriptionCreate {
  name: string;
  endpoint_url: string;
  http_method?: HttpMethod;
  headers?: Record<string, string>;
  auth_type?: AuthType;
  secret_key?: string;
  auth_user?: string;
  filter_expression?: Array<{ field: string; op: string; value: unknown }>;
  event_codes?: string[];
  transform_template?: Record<string, string>;
  max_retries?: number;
  timeout_seconds?: number;
}

export interface DeliveryAttempt {
  id: string;
  event_log_id: string;
  subscription_id: string;
  attempt_no: number;
  response_status?: number;
  response_body?: string;
  error_message?: string;
  attempted_at?: string;
  next_retry_at?: string;
  status: DeliveryStatus;
  duration_ms?: number;
  created_at: string;
}

export interface InboundEndpoint {
  id: string;
  name: string;
  endpoint_path: string;
  method: HttpMethod;
  auth_type: AuthType;
  handler_module?: string;
  description?: string;
  active_flag: boolean;
  last_received_at?: string;
  receive_count: number;
  created_at: string;
}

export interface WebhookDashboard {
  total_events_24h: number;
  total_events_7d: number;
  total_deliveries_24h: number;
  delivery_success_rate_24h: number;
  pending_deliveries: number;
  dead_letter_count: number;
  active_subscriptions: number;
  top_events: Array<{ event_code: string; count: number }>;
  recent_failures: Array<Record<string, unknown>>;
}

export interface WebhookReport {
  delivery_success_rate: number;
  total_delivered: number;
  total_failed: number;
  total_dead_letter: number;
  avg_retry_count: number;
  top_event_producers: Array<{ module: string; count: number }>;
  failure_reasons: Array<{ http_status: number; count: number }>;
  subscription_latency: Array<{ subscription_id: string; avg_latency_ms: number }>;
}

const BASE = "/api/v1/webhooks";

export async function getDashboard(): Promise<WebhookDashboard> {
  const res = await apiClient.get<WebhookDashboard>(`${BASE}/dashboard`);
  return res.data;
}

export async function listEventDefinitions(module?: string): Promise<EventDefinition[]> {
  const res = await apiClient.get<EventDefinition[]>(`${BASE}/event-definitions`, { params: { module } });
  return res.data;
}

export async function seedEventDefinitions(): Promise<void> {
  await apiClient.post(`${BASE}/event-definitions/seed`);
}

export async function listEvents(params?: {
  event_code?: string; module?: string; reference_id?: string; limit?: number;
}): Promise<EventLog[]> {
  const res = await apiClient.get<EventLog[]>(`${BASE}/events`, { params });
  return res.data;
}

export async function publishEvent(data: {
  event_code: string; payload: Record<string, unknown>;
  reference_type?: string; reference_id?: string;
}): Promise<EventLog> {
  const res = await apiClient.post<EventLog>(`${BASE}/events/publish`, data);
  return res.data;
}

export async function replayEvents(data: {
  event_ids?: string[]; event_code?: string;
  from_dt?: string; to_dt?: string; subscription_ids?: string[];
}): Promise<{ detail: string }> {
  const res = await apiClient.post<{ detail: string }>(`${BASE}/events/replay`, data);
  return res.data;
}

export async function listSubscriptions(activeOnly = false): Promise<Subscription[]> {
  const res = await apiClient.get<Subscription[]>(`${BASE}/subscriptions`, { params: { active_only: activeOnly } });
  return res.data;
}

export async function createSubscription(data: SubscriptionCreate): Promise<Subscription> {
  const res = await apiClient.post<Subscription>(`${BASE}/subscriptions`, data);
  return res.data;
}

export async function updateSubscription(id: string, data: Partial<SubscriptionCreate & { active_flag: boolean }>): Promise<Subscription> {
  const res = await apiClient.patch<Subscription>(`${BASE}/subscriptions/${id}`, data);
  return res.data;
}

export async function deleteSubscription(id: string): Promise<void> {
  await apiClient.delete(`${BASE}/subscriptions/${id}`);
}

export async function testSubscription(id: string): Promise<{ success: boolean; status_code?: number; response?: string; error?: string }> {
  const res = await apiClient.post(`${BASE}/subscriptions/${id}/test`);
  return res.data;
}

export async function listDeliveries(params?: {
  subscription_id?: string; delivery_status?: DeliveryStatus; limit?: number;
}): Promise<DeliveryAttempt[]> {
  const res = await apiClient.get<DeliveryAttempt[]>(`${BASE}/deliveries`, { params });
  return res.data;
}

export async function listDeadLetter(limit = 100): Promise<DeliveryAttempt[]> {
  const res = await apiClient.get<DeliveryAttempt[]>(`${BASE}/deliveries/dead-letter`, { params: { limit } });
  return res.data;
}

export async function retryDelivery(id: string): Promise<DeliveryAttempt> {
  const res = await apiClient.post<DeliveryAttempt>(`${BASE}/deliveries/${id}/retry`);
  return res.data;
}

export async function processPending(): Promise<{ processed: number; retried: number }> {
  const res = await apiClient.post<{ processed: number; retried: number }>(`${BASE}/deliveries/process-pending`);
  return res.data;
}

export async function listInboundEndpoints(): Promise<InboundEndpoint[]> {
  const res = await apiClient.get<InboundEndpoint[]>(`${BASE}/inbound-endpoints`);
  return res.data;
}

export async function createInboundEndpoint(data: Partial<InboundEndpoint>): Promise<InboundEndpoint> {
  const res = await apiClient.post<InboundEndpoint>(`${BASE}/inbound-endpoints`, data);
  return res.data;
}

export async function getReport(days = 7): Promise<WebhookReport> {
  const res = await apiClient.get<WebhookReport>(`${BASE}/reports`, { params: { days } });
  return res.data;
}

export async function getHealthMonitor() {
  const res = await apiClient.get(`${BASE}/ai/health-monitor`);
  return res.data;
}

export async function getPayloadAdvisor() {
  const res = await apiClient.get(`${BASE}/ai/payload-advisor`);
  return res.data;
}

export async function getAnomalyDetector() {
  const res = await apiClient.get(`${BASE}/ai/anomaly-detector`);
  return res.data;
}

export const STATUS_BADGE: Record<DeliveryStatus, string> = {
  pending: "bg-yellow-100 text-yellow-800",
  sent: "bg-green-100 text-green-800",
  failed: "bg-red-100 text-red-800",
  retrying: "bg-blue-100 text-blue-800",
  dead_letter: "bg-gray-900 text-white",
};
